File: python/src/onchaindiligence/agent_evidence/schema.py
"""Offline JSON Schema loading and validation."""

from __future__ import annotations

import json
from functools import cache, lru_cache
from importlib.resources import files
from typing import Any

from jsonschema import Draft202012Validator, FormatChecker
from referencing import Registry, Resource
from referencing.jsonschema import DRAFT202012

from .errors import SchemaValidationError

SCHEMA_BASE = "https://onchaindiligence.com/schemas/agent-evidence/v0/"
SCHEMA_NAMES = (
    "common.schema.json",
    "attestation-key.schema.json",
    "dsse-envelope.schema.json",
    "proof.schema.json",
    "record.schema.json",
    "bundle-payload.schema.json",
    "portable-file.schema.json",
)


@lru_cache(maxsize=1)
def _schemas() -> dict[str, dict[str, Any]]:
    directory = files("onchaindiligence.agent_evidence").joinpath("schemas")
    loaded: dict[str, dict[str, Any]] = {}
    for name in SCHEMA_NAMES:
        raw = directory.joinpath(name).read_text(encoding="utf-8")
        value = json.loads(raw)
        if not isinstance(value, dict):
            raise RuntimeError(f"packaged schema {name} is not an object")
        loaded[name] = value
    return loaded


@lru_cache(maxsize=1)
def _registry() -> Registry[Any]:
    resources = [
        (
            str(schema["$id"]),
            Resource.from_contents(schema, default_specification=DRAFT202012),
        )
        for schema in _schemas().values()
    ]
    return Registry().with_resources(resources)


@cache
def validator(name: str) -> Draft202012Validator:
    """Return a validator whose references are resolved only from package data."""

    try:
        schema = _schemas()[name]
    except KeyError as exc:
        raise ValueError(f"unknown Agent Evidence schema: {name}") from exc
    Draft202012Validator.check_schema(schema)
    return Draft202012Validator(schema, registry=_registry(), format_checker=FormatChecker())


def validate_document(name: str, value: Any) -> None:
    """Validate a document and raise one deterministic, path-aware error."""

    errors = sorted(validator(name).iter_errors(value), key=lambda item: list(item.absolute_path))
    if not errors:
        return
    error = errors[0]
    path = "/" + "/".join(str(part) for part in error.absolute_path)
    raise SchemaValidationError(f"{name}{path}: {error.message}")

File: packages/agent-evidence/src/schema.ts
import { readFileSync } from 'node:fs'
import Ajv2020Import, { type ErrorObject, type ValidateFunction } from 'ajv/dist/2020.js'
import addFormatsImport from 'ajv-formats'
import { SchemaValidationError } from './errors.js'

export const SCHEMA_BASE = 'https://onchaindiligence.com/schemas/agent-evidence/v0/'
export const SCHEMA_NAMES = [
  'common.schema.json',
  'attestation-key.schema.json',
  'dsse-envelope.schema.json',
  'proof.schema.json',
  'record.schema.json',
  'bundle-payload.schema.json',
  'portable-file.schema.json',
] as const

interface AjvRuntime {
  addSchema(schema: unknown): void
  getSchema(id: string): ValidateFunction | undefined
}

const Ajv2020 = Ajv2020Import as unknown as new (options: Record<string, unknown>) => AjvRuntime
const addFormats = addFormatsImport as unknown as (instance: AjvRuntime) => void
const ajv = new Ajv2020({ allErrors: true, strict: true })
addFormats(ajv)
for (const name of SCHEMA_NAMES) {
  const bytes = readFileSync(new URL(`../schemas/${name}`, import.meta.url), 'utf8')
  ajv.addSchema(JSON.parse(bytes))
}

function validator(name: string): ValidateFunction {
  if (!(SCHEMA_NAMES as readonly string[]).includes(name)) {
    throw new RangeError(`unknown Agent Evidence schema: ${name}`)
  }
  const validate = ajv.getSchema(`${SCHEMA_BASE}${name}`)
  if (!validate) throw new Error(`packaged Agent Evidence schema did not compile: ${name}`)
  return validate
}

function errorSortKey(error: ErrorObject): string {
  return `${error.instancePath}\u0000${error.schemaPath}\u0000${error.keyword}`
}

export function validateDocument(name: string, value: unknown): void {
  const validate = validator(name)
  if (validate(value)) return
  const errors = [...(validate.errors ?? [])].sort((left, right) => errorSortKey(left).localeCompare(errorSortKey(right)))
  const error = errors[0]
  if (!error) throw new SchemaValidationError(`${name}/: schema validation failed`)
  throw new SchemaValidationError(`${name}${error.instancePath || '/'}: ${error.message ?? error.keyword}`)
}
